universe/Universe.cpp

```diff
@@ -1660,6 +1660,17 @@ void Universe::ForgetKnownObject(int empire_id, int object_id) {
     objects.Remove(object_id);
 }
 
+void Universe::GrantShipDesignKnowledge(int empire_id, int object_id) {
+    std::shared_ptr<const Ship> ship = GetShip(object_id);
+    if (!ship) return;
+    int design_id = ship->DesignID();
+    if (design_id == ShipDesign::INVALID_DESIGN_ID) {
+        ErrorLogger() << "GrantShipDesignKnowledge got invalid design id for ship with id " << object_id;
+    } else {
+        m_empire_known_ship_design_ids[empire_id].insert(design_id);
```

Comment:
OK, if you prefer to duplicate a check for a more localized error message, that's reasonable.

I'd not call it HandleShipDesignKnowledge though, as the "handle" implies a response to an event, or handling all ship design knowledge stuff together, rather than doing it for a specific empire and ship id. So go with SetEmpireKnowledgeOfDesignOfShip please.